tests/test_filesystems.py

```diff
@@ -0,0 +1,344 @@
+        with synapse_errors("foo"):
+            raise SynapseHTTPError("already exists")
+    with pytest.raises(SynapseHTTPError):
+        with synapse_errors("foo"):
+            raise SynapseHTTPError("something else")
+    with pytest.warns(UserWarning):
+        with synapse_errors("foo"):
+            raise SynapseHTTPError("File size must be at least one byte")
+
+
+def test_that_a_remote_file_without_a_close_on_callable_can_be_closed():
+    with TemporaryFile() as temp_file:
+        remote_file = RemoteFile(temp_file, "foo", "w", on_close=None)
+        remote_file.close()
+
+
+def test_that_staging_a_local_file_creates_a_copy(get_data):
+    path = get_data("test.txt")
+    local_fs = open_fs(f"osfs://{path.parent}")
+    with TemporaryDirectory() as tmp_dir_name:
+        tmp_dir_path = Path(tmp_dir_name)
+        target_path = tmp_dir_path / "test.txt"
+        assert not target_path.exists()
+        target_file = target_path.open("wb")
+        local_fs.download(path.name, target_file)
+        target_file.close()
+        assert target_path.exists()
+
+
+@pytest.mark.integration
+def test_that_staging_a_synapse_file_creates_a_copy(mocker):
+    mocked_synapse = mocker.patch("synapseclient.Synapse")
+    synapse_fs = open_fs("syn://syn50545516")
+    with TemporaryDirectory() as tmp_dir_name:
+        tmp_dir_path = Path(tmp_dir_name)
+        tmp_file_path = tmp_dir_path / "mocked.txt"
+        with tmp_file_path.open("w") as tmp_file:
+            tmp_file.write("foobar")
+        mocked_entity = mocker.PropertyMock(tmp_file_path)
+        mocked_synapse.get.return_value = mocked_entity
+        target_path = tmp_dir_path / "test.txt"
+        assert not target_path.exists()
+        target_file = target_path.open("wb")
+        synapse_fs.download("test.txt", target_file)
+        target_file.close()
+        assert target_path.exists()
+
+
+@pytest.mark.integration
+class TestSynapseFS(FSTestCases, unittest.TestCase):
+    TEST_ROOT_PARENT = "syn50555278"
+
+    @classmethod
+    def setUpClass(cls):
+        cls.auth_token = os.environ.get("SYNAPSE_AUTH_TOKEN")
+        if cls.auth_token is None:
+            cls.skipTest(cls, "'SYNAPSE_AUTH_TOKEN' not set in environment.")
+        cls.synapse = Synapse()
+        cls.synapse.login(authToken=cls.auth_token)
+
+        session_name = pytest.UUID  # type: ignore
+        cls.test_root = Folder(session_name, parent=cls.TEST_ROOT_PARENT)
+        cls.test_root = cls.synapse.store(cls.test_root)
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.synapse.delete(cls.test_root)
+
+    def make_fs(self):
+        folder_name = self.id()
+        self.folder = Folder(folder_name, parent=self.test_root)
+        self.folder = self.synapse.store(self.folder)
+        synapse_fs = SynapseFS(self.folder.id, self.auth_token)
+        return synapse_fs
+
+    def destroy_fs(self, fs):
+        fs.close()
+        self.synapse.delete(self.folder)
```

Comment:
I'm assuming that `make_fs` is what assigns the `self.fs`?  Or does this come from the extension of the `FSTestCases`.

If I'm understanding correctly, you are instantiating a FS that has the root folder `self.folder.id` - Is that correct?

Reply:
`make_fs()` is called by `setUp()` in the base test class [here](https://github.com/PyFilesystem/pyfilesystem2/blob/8ed9dc495d8ba2f83fbb2a1145d34d92e13644be/fs/test.py#L268-L269), which in turn is called automatically by `unittest` for each test method. 

And you're correct on the second point. I'm creating a new Synapse folder to act as the root for each individual test. It's part of why this test suite takes so long, but it ensures that the tests don't run into concurrency problems.